### PR TITLE
[ONNX] Produce comprehensive assertion errors for quantized outputs

### LINKED ISSUE #83038: [ONNX] Produce comprehensive assertion errors for quantized outputs
### 🚀 The feature, motivation and pitch

Currently in `_compare_ort_pytorch_outputs`, quantized tensors' zero points and scales are flattened with the int8 representation for comparison. When int8 tensors do not match, one has to print out the scales and zero points to find that the mismatch is caused by different output scales and zero points. At the same time we do not have a clear indication of whether a tensor being compared is quantized or not.

It would be helpful to handle quantized tensor comparison differently to produce clear error messages when a discrepancy is found in test.

cc @bowen

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jazzysoggy**
Fixes #83038

Currently ``_compare_ort_pytorch_outputs`` does not produce clearer error messages for differences in the zero point or scale of the two outputs. It also does not produce a clear error message for whether both are quantized.

This pull request adds assertions to output whether the scales and zero points have differences, and whether each individual output is quantized.

### CODE REVIEW

**Comment by justinchuby** [comment]
Thanks for the PR @jazzysoggy . ort_outs and pt_outs are two lists of np arrays. I suggest the following:

In the for loop, if both outputs are of dtype uint8 and we see an assertion error (under `except AssertionError as e:`), warning that the output can be because of a difference in quantization scale and zero_point, then raise.

You may check the tests `onnx/test` in CI to ensure the change works

**Review by justinchuby — CHANGES_REQUESTED** [review]
Please see comment and test failures

**Comment by jazzysoggy** [comment]
> Please see comment and test failures

I can’t see the comments or the changes requested

File: torch/onnx/verification.py [hunk 1]
```diff
@@ -199,6 +199,17 @@ def _compare_ort_pytorch_outputs(
                         f"within acceptable range {acceptable_error_percentage}."
                     )
                     continue
+                else:
+                    if ort_out.is_quantized and pt_out.is_quantized:
```
Comment (justinchuby) [inline torch/onnx/verification.py:203]: @jazzysoggy  ort_out is a bumpy array and doesn’t have the is_quantized attribute. I’d check it’s dtype and infer that it is a quantized tensor.

Comment (jazzysoggy) [reply torch/onnx/verification.py:203]: Used dtype to infer quantization by checking whether the dtype is uint8 or int8

**Comment by justinchuby** [comment]
Hi, I suggest testing the changes locally to make sure they work. In particular, I don't think the methods you are calling on the tensors exist, as they are essentially np arrays. I will give this issue a little more thought on how the error experience should be. For now I think we should add a hint when uint8 or int8 formats are detected to tell users they may be dealing with quantized tensors.

**Comment by jazzysoggy** [comment]
I was wondering how would I find the scale and zero_point without those two functions for the ONNX tensor. I tried searching documentation, but I couldn’t seem to find it.

Also, if the PyTorch tensor is quantized, shouldn’t it have those two methods.

**Comment by jazzysoggy** [comment]
Can I close this PR and open a new one? It seems like my attempt to sync the branch broke the entire PR.

**Comment by justinchuby** [comment]
Sure thing. Thanks

**Pull request closed without merging**